mesonbuild/modules/i18n.py

```diff
@@ -92,7 +92,7 @@ def merge_file(self, state, args, kwargs):
         outputs = mesonlib.substitute_values([output], values)
         output = outputs[0]
 
-        ct = build.CustomTarget(output + '_' + state.subdir + '_merge', state.subdir, state.subproject, kwargs)
+        ct = build.CustomTarget(output + '_' + state.subdir.replace('/', '__') + '_merge', state.subdir, state.subproject, kwargs)
```

Comment:
I tested this on Windows and found an issue: On Windows, this doesn't fix the `Target "xxx" has a path separator in its name.` warnings because of path separator (backslash).
I think it would be better to replace `/` with `os.sep` (= backslash on Windows, slash on POSIX systems).

```python
from os import path, sep
# ...
        ct = build.CustomTarget(output + '_' + state.subdir.replace(sep, '__') + '_merge', state.subdir, state.subproject, kwargs)
```

Reply:
Or, always replace both `\` and `/`. Note that elsewhere we use `@@` as the replacement character instead of `__`, so this PR should do the same.

Reply:
To suppress the warnings, another solution is to skip the "path separator check" only when `CustomTarget` object is created in `i18n.merge_file()`:
```diff
--- a/mesonbuild/build.py
+++ b/mesonbuild/build.py
@@ -310,8 +310,8 @@ class EnvironmentVariables:
         return env
 
 class Target:
-    def __init__(self, name, subdir, subproject, build_by_default):
-        if has_path_sep(name):
+    def __init__(self, name, subdir, subproject, build_by_default, check_path_sep=True):
+        if check_path_sep and has_path_sep(name):
             # Fix failing test 53 when this becomes an error.
             mlog.warning('''Target "%s" has a path separator in its name.
 This is not supported, it can cause unexpected failures and will become
@@ -1720,8 +1720,8 @@ class CustomTarget(Target):
         'console',
     ])
 
-    def __init__(self, name, subdir, subproject, kwargs, absolute_paths=False):
-        super().__init__(name, subdir, subproject, False)
+    def __init__(self, name, subdir, subproject, kwargs, absolute_paths=False, check_path_sep=True):
+        super().__init__(name, subdir, subproject, False, check_path_sep)
         self.dependencies = []
         self.extra_depends = []
         self.depend_files = [] # Files that this target depends on but are not on the command line.
--- a/mesonbuild/modules/i18n.py
+++ b/mesonbuild/modules/i18n.py
@@ -92,7 +92,7 @@ class I18nModule(ExtensionModule):
         outputs = mesonlib.substitute_values([output], values)
         output = outputs[0]
 
-        ct = build.CustomTarget(output + '_' + state.subdir + '_merge', state.subdir, state.subproject, kwargs)
+        ct = build.CustomTarget(output + '_' + state.subdir + '_merge', state.subdir, state.subproject, kwargs, check_path_sep=False)
         return ModuleReturnValue(ct, [ct])
 
     @FeatureNewKwargs('i18n.gettext', '0.37.0', ['preset'])
```
This doesn't change build messages ("Generating xxx_merge with a custom command."). In that case, we may need to add path separator checks in `merge_file()` (for `kwargs['input']` and `kwargs['output']`).

I'm not sure what is the best solution.

Reply:
Alrighty, just adjusted and pushed again. Replaces either a `/` or `\` with `@`.